projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs

```diff
@@ -68,25 +68,25 @@ internal class SocketFrameHandler : IFrameHandler
         private TimeSpan _writeableStateTimeout = TimeSpan.FromSeconds(30);
         private int _writeableStateTimeoutMicroSeconds;
         private readonly ITcpClient _socket;
-        private readonly Stream _reader;
-        private readonly Stream _writer;
-        private readonly ChannelWriter<ReadOnlyMemory<byte>> _channelWriter;
-        private readonly ChannelReader<ReadOnlyMemory<byte>> _channelReader;
+        private readonly ChannelWriter<OutgoingFrame> _channelWriter;
+        private readonly ChannelReader<OutgoingFrame> _channelReader;
+        public PipeReader PipeReader { get; private set; }
+
+        private readonly PipeWriter _pipeWriter;
         private readonly Task _writerTask;
         private readonly object _semaphore = new object();
         private readonly byte[] _frameHeaderBuffer;
+        private readonly IMeasuredDuplexPipe _pipe;
         private bool _closed;
 
-        public SocketFrameHandler(AmqpTcpEndpoint endpoint,
-            Func<AddressFamily, ITcpClient> socketFactory,
-            TimeSpan connectionTimeout, TimeSpan readTimeout, TimeSpan writeTimeout)
+        public SocketFrameHandler(AmqpTcpEndpoint endpoint, Func<AddressFamily, ITcpClient> socketFactory, TimeSpan connectionTimeout, TimeSpan readTimeout, TimeSpan writeTimeout)
         {
             Endpoint = endpoint;
-            _frameHeaderBuffer = new byte[6];
-            var channel = Channel.CreateUnbounded<ReadOnlyMemory<byte>>(
+            _frameHeaderBuffer = new byte[8];
+            var channel = Channel.CreateUnbounded<OutgoingFrame>(
                 new UnboundedChannelOptions
                 {
-                    AllowSynchronousContinuations = false,
+                    AllowSynchronousContinuations = true,
```

Comment:
What's the resulting change for this?
Does publish now possibly write to the pipe/socket? If so is this what we want?

Reply:
Publish will write to the channel, but it might run synchronously if the WriteLoop completes synchronously, for example Pipeline had enough empty buffer to receive the data. This is what we want to avoid the async overhead for synchronous completions since the Pipelines take care of async.

Reply:
But this also means that the thread that called publish might be blocked for a longer time, isn't-it?